src/cmd/ksh93/sh/io.c

```diff
@@ -1319,6 +1319,7 @@ int	sh_redirect(Shell_t *shp,struct ionod *iop, int flag)
 				if(flag==SH_SHOWME)
 					goto traceit;
 				fd=sh_chkopen(fname);
+				fd=sh_iomovefd(shp, fd);
```

Comment:
This breaks the `directoryfd` test. Without this change the added unit tests still succeed. Presumably this issue has already been fixed in some other fashion. We could merge just the added unit tests. Want to take a closer look, @siteshwar?